htdoc/htnotify.1.in
.TH htnotify 1 "21 July 1997"
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)
.SH NAME
htnotify \- sends email notifications about out-dated web pages discovered by htmerge
.SH SYNOPSIS
.B htnotify
.I "[options]"
.SH "DESCRIPTION"
.I Htnotify
scans the document database created by
.I htmerge
and sends an email message for every page that is out of date. Please
have a look at the ht://Dig notification manual for instructions on how to
set up this service. 
.SH OPTIONS
.TP
.B \-b \fIdatabase\fR
Specifies an alternative database than what is specified in
the configuration file. 
.TP
.B \-c \fIconfigfile\fR
Use the specified
.I configfile
instead of the default. 
.B \-v
Verbose mode. This increases the verbosity of the program.
Used once will display a log of what email messages were
sent. Used more than once will display information about
each document that has email notification set.
.SH "FILES"
.TP
.B @DEFAULT_CONFIG_FILE@
The default configuration file.
.SH "SEE ALSO"
Please refer to the HTML pages (in the htdig-doc package)
.B (THIS MUST BE CUSTOMISED...)
.B /usr/share/doc/htdig-doc/html/index.html
and the manual pages
.I htdig(1)
and 
.I htmerge(1)
for a detailed description of ht://Dig and its commands.
.SH AUTHOR
This manual page was written by Christian Schwarz, modified by
Stijn de Bekker, based on the HTML documentation of ht://Dig.


ChangeLog
Wed Feb 23 16:53:40 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/defaults.cc (backlink_factor, external_parser,
	local_default_doc, local_urls, local_urls_only, local_user_urls):
	Add some updates from 3.1.5's attrs.html.
	* htdoc/attrs.html, htdoc/cf_byname.html, htdoc/cf_byprog.html:
	Regenerate using cf_generate.pl.

Wed Feb 23 15:11:51 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	[ Improve htsearch's HTML 4.0 compliance ]
	* htsearch/TemplateList.cc (createFromString): Use file name rather
	than internal name to select builtin-* templates, use $&(TITLE) and
	$&(URL) in templates and quote HTML tag parameters.
	* installdir/long.html, installdir/short.html: Use $&(TITLE) and
	$&(URL) in templates and quote HTML tag parameters.
	* htsearch/Display.cc (setVariables): quote all HTML tag parameters
	in generated select lists.
	* installdir/footer.html, installdir/header.html,
	installdir/nomatch.html, installdir/search.html,
	installdir/syntax.html, installdir/wrapper.html:
	Use $&(var) where appropriate, and quote HTML tag parameters.
	* installdir/htdig.conf: quote all HTML tag parameters.

Wed Feb 23 13:40:27 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/URL.h (encodeURL): Change list of valid characters to
	include only unreserved ones.
	* htcommon/cgi.cc (init): Allow "&" and ";" as input param. separators.
	* htsearch/Display.cc (createURL): Encode each parameter separately,
	using new unreserved list, before piecing together query string, to
	allow characters like "?=&" within parameters to be encoded.

Wed Feb 23 13:22:29 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/URL.cc (ServerAlias): Fix server_aliases processing to prevent
	infinite loop (as for local_urls in PR#688).

Wed Feb 23 12:49:52 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/HtDateTime.h, htlib/HtDateTime.cc: change Httimegm() method
	to HtTimeGM(), to avoid conflict with Httimegm() C function, so we
	don't need "::" override, for Mac OS X.
	* htlib/htString.h, htlib/String.cc: change write() method to
	Write(), to avoid conflict with write() function, so we don't need
	"::" override, for Mac OS X.

Wed Feb 23 12:17:46 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/Configuration.cc(Read): Fixed to allow final line without
	terminating newline character, rather than ignoring it.

Wed Feb 23 12:01:01 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/Retriever.cc (GetLocal, GetLocalUser): Add URL-decoding
	enhancements to local_urls, local_default_urls & local_default_doc,
	to allow hex encoding of special characters.

Wed Feb 23 19:14:29 2000  Loic Dachary  <loic@ceic.com>

 	* htcommon/conf_parser.cxx: regenerated from conf_parser.yxx

Wed Feb 23 19:04:16 2000  Loic Dachary  <loic@ceic.com>

	* test/test_functions.in: inconditionaly remove existing test/var
	  directory before runing tests to prevent accidents.

	* htcommon/URL.cc (URL): fixed String->char warning

	* htcommon/defaults.cc (wordlist_compress): defaults to true

Tue Feb 22 17:09:10 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/HTML.cc(parse, do_tag): Fix handling of <img alt=...> text
	and parsing of words in meta tags, to to proper word separation.
	* htlib/HtWordType.h, htlib/HtWordType.cc: Add HtWordToken() function,
	to replace strtok() in HTML parser.

Tue Feb 22 16:21:25 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htlib/URL.cc (ctor, normalizePath): Fix PR#779, to handle relative
	URLs correctly when there's a trailing ".." or leading "//".

Tue Feb 22 14:09:26 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htdig/Document.cc (RetrieveLocal): Handle common extensions for
	text/plain, application/pdf & application/postscript.

Mon Feb 21 17:25:21 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* contrib/htdig-3.2.0.spec: Fixed %post script to add more
	descriptive entries in htdig.conf, made cron script a config file,
	updated to 3.2.0b2.

	* contrib/conv_doc.pl, contrib/parse_doc.pl: Added comments to show
	Warren Jones's updates in change history.

Mon Feb 21 17:09:13 2000  Gilles Detillieux  <grdetil@scrc.umanitoba.ca>

	* htcommon/HtConfiguration.h, htcommon/conf_parser.yxx,
	htlib/Configuration.h, htlib/Configuration.cc: split Add() method
	into Add() and AddParsed(), so that only config attributes get parsed.
	Use AddParsed() only in Read() and Defaults().

Sat Feb 19 04:56:47 2000  Geoff Hutchison  <ghutchis@wso.williams.edu>

	* htnet/Connection.h, htnet/Connection.cc: Renamed methods with
	capitals to remove the need to use ::-escaped library calls.

Fri Feb 18 14:40:50 2000  Loic Dachary  <loic@ceic.com>

	* test/conf/access.conf.in: removed cookies. Not used and some
	  httpd are not compiled with usertrack.

Wed Feb 16 12:15:08 2000  Vadim Chekan  <vadim@etc.lviv.ua>

	* htcommon/Makefile.am replaced conf.tab.cc.h by conf_parser.h in
	noinst_HEADERS

	* htcommon/conf_parser.yxx,conf_parser.lxx,HtConfiguration.cc,
	HtConfiguration.h: added copyright and Id:

	* htcommon/cgi.cc(init): fixed bug: array must be free by 
	delete [] buf, not just delete buf;

Tue Feb 15 23:16:14 2000  Geoff Hutchison  <ghutchis@wso.williams.edu>

	* htnet/HtHTTP.cc (isParsable): Remove application/pdf as a
	default type--it is now handled through the ExternalParser
	interface if at all.

	* htcommon/defaults.cc: Remove pdf_parser attribute.

	* htdoc/attrs.html, htdoc/cf_byname.html, htdoc/cf_byprog.html:
	Regenerate using cf_generate.pl.

	* htdig/Document.cc (getParsable): Remove PDF once and for all
	(hopefully).

	* htdig/ExternalParser.cc (parse): Ditto.

	* configure.in: Remove check for PDF_PARSER.

	* configure: Regenerate using autoconf

	* htdig/Makefile.am: Remove PDF.cc and PDF.h.

	* Makefile.in, */Makefile.in: Regenerate using automake --foreign

Tue Feb 15 12:02:39 EET 2000  Vadim Chekan  <vadim@etc.lviv.ua>

	* htcommon/HtConfiguration.cc,HtConfiguration.h: fixed bug discovered

src/tools/htnotify.cc
//
// htnotify.cc
//
// htnotify: Check through databases and look for notify META information
//           Send e-mail to addresses mentioned in documents if the doc
//           has "expired"
//
// Part of the ht://Dig package   <http://www.htdig.org/>
// Copyright (c) 1995-2004 The ht://Dig Group
// For copyright details, see the file COPYING in your distribution
// or the GNU Library General Public License (LGPL) version 2 or later
// <http://www.gnu.org/copyleft/lgpl.html>
//
// $Id: htnotify.cc,v 1.1.2.1 2006/09/25 23:51:20 aarnone Exp $
//

#ifdef HAVE_CONFIG_H
#include "htconfig.h"
#endif /* HAVE_CONFIG_H */

#include "HtConfiguration.h"
#include "Dictionary.h"
#include "DocumentDB.h"
#include "DocumentRef.h"
#include "defaults.h"
#include "HtURLCodec.h"

#include <stdlib.h>

#ifndef _MSC_VER /* _WIN32 */
#include <unistd.h>
#endif
#include <errno.h>

#ifdef HAVE_STD
#include <fstream>
#ifdef HAVE_NAMESPACES
using namespace std;
#endif
#else
#include <fstream.h>
#endif /* HAVE_STD */

#include <time.h>
#include <stdio.h>
#include <ctype.h>

// If we have this, we probably want it.
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#elif HAVE_GETOPT_LOCAL
#include <getopt_local.h>
#endif

// Declare a record for storing email/URL data for later retrieval
class EmailNotification : public Object
{
public:
    EmailNotification (char* date, char* email, char* url, char* subject);

    //
    //accessors
    //
    String getDate()    const { return date; }
    String getEmail()   const { return email; }
    String getUrl()     const { return url; }
    String getSubject() const { return subject; }

private:
    String date;
    String email;
    String url;
    String subject;
};

EmailNotification::EmailNotification (char* pDate, char* pEmail,
                                      char* pUrl,  char* pSubject)
{
    date    = pDate;
    email   = pEmail;
    url     = pUrl;
    if (!pSubject || !*pSubject)
    {
        subject = "page expired";
    }
    else
    {
        subject = pSubject;
    }
}

void htnotify(DocumentRef &);
void usage();
void readPreAndPostamble(void);
void add_notification(char *date, char *email, char *url, char *subject);
void send_notification(char *email, List * notifList);
void send_email(List * notifList, String& command, String& to,
		String& listText, int singleSubject);
int parse_date(char *date, int &year, int &month, int &day);


int	verbose = 0;
int	sendEmail = 1;

//
// This variable is used to hold today's date.  It is global for
// efficiency reasons since computing it is a relatively expensive
// operation
//
struct tm	*today;

//
// This structure holds the set of email notifications requiring
// sending. It is indexed by email address of recipients, and
// each entry is a List of EmailNotification objects.
//
Dictionary  * allNotifications;

//
// These strings holds the preamble/postamble text used in
// email messages.
//
String preambleText;
String postambleText;

//{{{  main
//*****************************************************************************
// int main(int ac, char **av)
//
int main(int ac, char **av)
{
    int			c;
    extern char		*optarg;
    String		base;
    String		configFile = DEFAULT_CONFIG_FILE;

    while ((c = getopt(ac, av, "nvb:c:")) != -1)
    {
	switch (c)
	{
	case 'b':
	    base = optarg;
	    break;
	case 'c':
	    configFile = optarg;
	    break;
	case 'v':
	    verbose++;
	    break;
	case 'n':
	    verbose++;
	    sendEmail = 0;
	    break;
	case '?':
	    usage();
	    break;
	}
    }

	HtConfiguration* config= HtConfiguration::config();
    config->Defaults(&defaults[0]);
    config->Read(configFile);

    //
    // Check url_part_aliases and common_url_parts for
    // errors.
    String url_part_errors = HtURLCodec::instance()->ErrMsg();

    if (url_part_errors.length() != 0)
    {
      cerr << form("htnotify: Invalid url_part_aliases or common_url_parts: %s",
                   url_part_errors.get()) << endl;
      exit (1);
    }

    if (base.length())
    {
	config->Add("database_base", base);
    }

    String	doc_db = config->Find("doc_db");
    DocumentDB	docdb;

    docdb.Read(doc_db);
    List	*docs = docdb.DocIDs();

    //
    // Compute today's date
    //
    time_t	now = time(0);
    today = localtime(&now);

    readPreAndPostamble();

    //
    // Traverse all the known documents to check for notification requirements
    //
    allNotifications = new Dictionary();
    DocumentRef	*ref;
    IntObject		*id;
    docs->Start_Get();
    while ((id = (IntObject *) docs->Get_Next()))
    {
	ref = docdb[id->Value()];
	if (ref)
	    htnotify(*ref);
	delete ref;
    }
    delete docs;

    //
    // Iterate through the list of notifications
    //
    allNotifications->Start_Get();
    char * email;
    while ((email = (char *) allNotifications->Get_Next()))
    {
        List * notifList = (List *) allNotifications->Find (email);
        send_notification(email, notifList);
    }

    //
    // tidy up
    //
    docdb.Close();
    delete allNotifications;
    return 0;
}


//}}}  
//{{{  readPreAndPostamble
//*****************************************************************************
// void readPreAndPostamble(void)
//
void readPreAndPostamble(void)
{
	HtConfiguration* config= HtConfiguration::config();
    const char* prefixfile = config->Find("htnotify_prefix_file");
    const char* suffixfile = config->Find("htnotify_suffix_file");

    // define default preamble text - blank string
    preambleText = "";

    if (prefixfile != NULL && *prefixfile)
    {
        ifstream    in(prefixfile);
        char        buffer[1024];

        if (! in.bad())
        {
            while (! in.bad() && ! in.eof())
            {
                in.getline(buffer, sizeof(buffer));
                if (in.eof() && !*buffer)
                    break;
                preambleText << buffer << '\n';
            }
            in.close();
        }
    }

    // define default postamble text
    postambleText = "";
    postambleText << "Note: This message will be sent again if you do not change or\n";
    postambleText << "take away the notification of the above mentioned HTML page.\n";
    postambleText << "\n";
    postambleText << "Find out more about the notification service at\n\n";
    postambleText << "    http://www.htdig.org/meta.html\n\n";
    postambleText << "Cheers!\n\nht://Dig Notification Service\n";

    if (suffixfile != NULL && *suffixfile)
    {
        ifstream    in(suffixfile);
        char        buffer[1024];

        if (! in.bad())
        {
            postambleText = "";
            while (! in.bad() && ! in.eof())
            {
                in.getline(buffer, sizeof(buffer));
                if (in.eof() && !*buffer)
                    break;
                postambleText << buffer << '\n';
            }
            in.close();
        }
    }

    if (verbose > 1)
    {
        cout << "Preamble text:" << endl;
        cout << preambleText << endl << endl;
        cout << "Postamble text:" << endl;
        cout << postambleText << endl;
        cout << endl;
    }
}

//}}}  
//{{{  htnotify
//*****************************************************************************
// void htnotify(DocumentRef &ref)
//
void htnotify(DocumentRef &ref)
{
    char	*date = ref.DocNotification();
    char	*email = ref.DocEmail();

    if (date && *date && email && *email)
    {
	if (verbose > 2)
	{
	    cout << "Saw a date:" << endl;
	    cout << "Date:    " << date << endl;
	    cout << "URL:     " << ref.DocURL() << endl;
	    cout << "Subject: " << ref.DocSubject() << endl;
	    cout << "Email:   " << email << endl;
	    cout << endl;
	}

	int		month, day, year;
	if (!parse_date(date, year, month, day))
	{
	    // Parsing Failed
	    if (verbose > 2)
	    {
		cout << "Malformed date: " << date << endl;
	    }

	    add_notification(date, email, ref.DocURL(), "Malformed Date");
	    return;
	}

	year -= 1900;
	month--;

	//
	// Compare this date with today's date
	//
	if (year < today->tm_year ||
	    (year == today->tm_year && month < today->tm_mon) ||
	    (year == today->tm_year && month == today->tm_mon &&
	     day < today->tm_mday))
	{
	    //
	    // It seems that this date is either today or before
	    // today.  Send a notification
	    //
	    add_notification(date, email, ref.DocURL(), ref.DocSubject());
	}
	else
	{
	    // Page not yet expired
	    if (verbose > 2)
	    {
		cout << "htnotify: URL " << ref.DocURL()
		     << " (" << year+1900 << "-" << month+1
		     << "-" << day << ")" << endl;
	    }
	}
    }
}


//}}}  
//{{{  add_notification
//*****************************************************************************
// void add_notification(char *date, char *email, char *url, char *subject)
//
void add_notification(char *date, char *email, char *url, char *subject)
{

    List * list = (List *) allNotifications->Find (email);
    if (list == NULL)
    {   // here's a new recipient so add it
        list = new List();
        allNotifications->Add (email, list);
    }

    // now add the notification to the selected list
    EmailNotification* notif = new EmailNotification(date, email, url, subject);
    list->Add (notif);
}

//}}}  
//{{{  send_notification
//*****************************************************************************
// void send_notification(char * email, List * notifList)
//
void send_notification(char* email, List * notifList)
{
    String	command = SENDMAIL;
    command << " -t";

    String	em = email;
    String	to = "";
    char	*token = strtok(em.get(), " ,\t\r\n");
    while (token)
    {
	if (*token)
	{
	    if (to.length())
		to << ", ";
	    to << token;
	}
	token = strtok(0, " ,\t\r\n");
    }

// Before we use the email address string, we may want to sanitize it.
//    static char ok_chars[] = "abcdefghijklmnopqrstuvwxyz
//    ABCDEFGHIJKLMNOPQRSTUVWXYZ
//    1234567890_-.@/=+:%!, ";
//    char *cursor;          // cursor into email address 
//    for (cursor = to.get(); *(cursor += strspn(cursor, ok_chars));)
//      *cursor = '_'; // Set it to something harmless
    
    EmailNotification* notif = (EmailNotification*) notifList->Get_First();
    String firstSubject = notif->getSubject();
    int singleSubject = 1;

    //
    // scan to determine whether the same subject message is used throughout
    //
    notifList->Start_Get();
    notifList->Get_Next();
    // continue with the second item in the list
    while ((notif = (EmailNotification*) notifList->Get_Next()))
    {
        String current = notif->getSubject();
        if ( firstSubject != current )
        {
            singleSubject = 0;
            break;
        }
    }


    //
    // Aggregate the list text
    //
    String listText = "";
    notifList->Start_Get();
    while ((notif = (EmailNotification*) notifList->Get_Next()))
    {
        listText << notif->getUrl() << '\n';
        listText << "  expired " << notif->getDate() << "\n";
        if (! singleSubject)
        { listText << "  " << notif->getSubject() << '\n'; }
    }

    if (sendEmail)
    {
        send_email (notifList, command, to, listText, singleSubject);
    }
    else if (verbose)
    {   // just list the notifiable pages
        cout << endl;
        cout << "Notification required to " << to << endl;
        cout << listText;
    }
}


//}}}  
//{{{  send_email
//*****************************************************************************
// void send_email(List * notifList, String& command, String& to)
//
void send_email (List * notifList, String& command,
                 String& to, String& listText, int singleSubject)
{
	HtConfiguration* config= HtConfiguration::config();
    String from = "\"";
    from << config->Find("htnotify_webmaster") << "\" <"
         << config->Find("htnotify_sender") << ">";

    String replyto = config->Find("htnotify_replyto");

    if (verbose)
    {
        if (verbose > 1) { cout << endl; }

        cout << "From: " << from << endl;
        cout << "To: " << to << endl;

        if (verbose > 1) { cout << listText; }
    }

    FILE *fileptr;
    if ( (fileptr = popen(command.get(), "w")) != NULL )
    {
        EmailNotification* notif = (EmailNotification*) notifList->Get_First();
        String	out;
        out << "From: " << from << '\n';
        out << "To: " << to << '\n';
        if (replyto.length() > 0)
        { out << "Reply-To: " << replyto << '\n'; }

        if (singleSubject)
        {
            out << "Subject: " << notif->getSubject() << '\n';
        }
        else
        {
            out << "Subject: Web page expiry (" << notif->getSubject() << ", inter alia)\n";
        }

        out << '\n'; // this is the important header/body separator
        out << preambleText;
        out << listText;
        out << postambleText;
        out << '\n';
        fputs( out.get(), fileptr );
        pclose( fileptr );
    }
    else
    {
        perror( "popen" );
    }
}


//*****************************************************************************
// Display usage information for the htnotify program
//
void usage()
{
    cout << "usage: htnotify [-c configfile][-b db_base]\n";
    cout << "This program is part of ht://Dig " << VERSION << "\n\n";
    cout << "There can be any number or words.\n";
    cout << "Options:\n";
    cout << "\t-c configfile\n";
    cout << "\t\tUse the specified configuration file instead of the default.\n\n";
    cout << "\t-b db_base\n";
    cout << "\t\tSet the base path of the document database.\n";
    cout << "\t-v\n";
    cout << "\t\tIncrease the verbose level. Use two or three times for\n";
    cout << "\t\tmore output.\n";
    cout << "\t-n\n";
    cout << "\t\tDon't send any email, just list what has expired.\n";
    exit(0);
}


//}}}  
//{{{  parse_date
//*****************************************************************************
// Parse the notification date string from the user's document
//
int parse_date(char *date, int &year, int &month, int &day)
{
	HtConfiguration* config= HtConfiguration::config();
    int		mm = -1, dd = -1, yy = -1, t;
    String	scandate = date;

    for (char *s = scandate.get(); *s; s++)
	if (ispunct(*s))
	    *s = ' ';

    if (config->Boolean("iso_8601"))
    {
	// conf file specified ISO standard, so expect [yy]yy mm dd.
	sscanf(scandate.get(), "%d%d%d", &yy, &mm, &dd);
    }
    else
    {
	// Default to American standard when not specified in conf,
	// so expect mm dd [yy]yy.
	sscanf(scandate.get(), "%d%d%d", &mm, &dd, &yy);
	if (mm > 31 && dd <= 12 && yy <= 31)
	{
	    // probably got yyyy-mm-dd instead of mm/dd/yy
	    t = mm; mm = dd; dd = yy; yy = t;
	}
    }

    // OK, we took our best guess at the order the y, m & d should be.
    // Now let's see if we guessed wrong, and fix it.  This won't work
    // for ambiguous dates (e.g. 01/02/03), which must be given in the
    // expected format.
    if (dd > 31 && yy <= 31)
    {
	t = yy; yy = dd; dd = t;
    }
    if (mm > 31 && yy <= 31)
    {
	t = yy; yy = mm; mm = t;
    }
    if (mm > 12 && dd <= 12)
    {
	t = dd; dd = mm; mm = t;
    }
    if (yy < 0 || mm < 1 || mm > 12 || dd < 1 || dd > 31)
	return 0;		// Invalid date

    if (yy < 70)		// before UNIX Epoch
	yy += 2000;
    else if (yy < 1900)		// before computer age
	yy += 1900;
    if (verbose > 2)
	cout << "Date used (y-m-d): " << yy << '-' << mm << '-' << dd << endl;

    year = yy;
    month = mm;
    day = dd;

    return 1;
}


//}}}  
